### ISSUE TITLE
phone-gap/cordova support

### ISSUE THREAD

**Issue by roeycohen** [ant-media/Ant-Media-Server#286]
# Issue
failing to play stream from cordova (phone-gap) application.

### Short description
ant-media is failing to play a published stream.
while opening https://HOSTNAME:5443/WebRTCAppEE/player.html from the regular chrome browser, running the same code from the application doesn't work:
![error](https://i.imgur.com/JKYscqe.jpg)

### Environment

[] Operating system and version: android 6
[] Java version: 1.8
[] Red5 version: 1.4.0-SNAPSHOT-180630_1610

### Expected behavior
the video will play

### Actual behavior
the doesn't play

### Steps to reproduce
1. create a basic cordova app
2. insert <iframe width="350" height="700" src="https://conferencing-video.blossom-kc.com:5443/WebRTCAppEE/player.html" frameborder="0" allowfullscreen></iframe> inside www/index.html
3. run the app, fill stream name and click start.

**Comment by mekya** [comment]
Could you please try directly in chrome? 

If it still fails,  I've recommend updating the chrome if there is an upate for you. Because there was a bug in some chrome versions about playing webrtc streams. After I updated the chrome, it is fixed.

**Comment by roeycohen** [comment]
Hi @mekya, thanks for the fast reply :)
the phone chrome browser is up to date and the video does play in it.

maybe we can learning something from the log?
![console.log](https://i.imgur.com/hfEo4t3.png)

**Comment by mekya** [comment]
Actually I've realized an issue yesterday. Android can send the audio but cannot send the video to the server and encoder did not started properly.

I mean server is still waiting for the video stream to start all encoders but video is not being received and even if stream exists in the server, it does not play. Is there something similar for your case?

Btw, your snapshot is almost one month old. You should update it ;)

**Comment by roeycohen** [comment]
I only tried to play the video... 🤷‍♂️

the stream does exists on the server because i'm able to play it from my pc and the chrome browser itself on my phone.

where can i get the most recent version? i'll make the test is it.

**Comment by roeycohen** [comment]
tried giving the application all possible permissions without success :(

**Comment by mekya** [comment]
If you play with Chrome browser on your phone but not playing in cordova app. 
There should be a special case about cordova. 

I can send you the link to download by e-mail. 
I've your e-mail. We are old friend, aren't we? :)

**Comment by roeycohen** [comment]
i guess it is a special case in cordova... i'll to do my best figure it out, but webrtc isn't my specialty :)

my email starts with roeycohen...

**Comment by mekya** [comment]
I see, the webview version, used by cordova ,is the same version as in the chrome? Can we know this info?

**Comment by roeycohen** [comment]
i'll try to find out.

**Comment by roeycohen** [comment]
navigator.userAgent is 
_Mozilla/5.0 (Linux; Android 7.0; SM-G920F Build/NRD90M; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/67.0.3396.87 Mobile Safari/537.36_
seems quite recent.

also tried going over the webview setting here: https://developer.android.com/reference/android/webkit/WebSettings
didn't found anything that related to webrtc.

what's bothers me most, is that using jitsi webrtc does work for me without any modifications.
maybe i should try another webrtc vendor to try isolate the problem better.

**Comment by roeycohen** [comment]
tried using this https://appr.tc/r/anttest service and it worked.
this must be something with the stream sent from ant.media server... probably something small.
i can upload my test project to github or send you by email so you can see for yourself.

**Comment by roeycohen** [comment]
i had another idea, maybe it has something to do with codecs. is it possible to change output codec? i'm willing to test this.

**Comment by mekya** [comment]
I see, we are using somethings different than other vendors. We use webrtc.org software to send stream with some modifications. I think, it may be an issue about h264 constrained profile.
https://bugs.chromium.org/p/webrtc/issues/detail?id=8584

What is your Android OS version? 
May I try your code in my device?

Thank you so much.

**Comment by roeycohen** [comment]
Hi @mekya,
i've tried on two phones:
S6 + android 7.0
1 plus3 + android 8.0
i can send you the source code, or compile an apk for you, let me know what's your preference is.

**Comment by roeycohen** [comment]
Hi @mekya,
i've created a small project for you so you could check yourself:
https://github.com/roeycohen/ant-media-cordova-test/tree/master

**Comment by roeycohen** [comment]
for anyone reading this, we've ended up understanding that the problem is with android webview not supporting h264 codec.

**Comment by mekya** [comment]
Here is a link for that
https://groups.google.com/a/chromium.org/forum/#!topic/android-webview-dev/hVZdHK6MKfI

Let me add something,  
actually there is a problem in android webview's h264 support in WebRTC.

**Comment by roeycohen** [comment]
Hi @meyka, just to be 100% sure, I'll try to make a test with jitsi with h264.

**Comment by mekya** [comment]
Yeah h264 is supported in webview but it is not supported in webrtc. 
Take a look at this message and above of it in the same thread 
https://groups.google.com/a/chromium.org/d/msg/android-webview-dev/hVZdHK6MKfI/cT-oSS--AgAJ

Sure please test it and let me know the results. If I've missed something, let it be out :)

**Comment by roeycohen** [comment]
Hmm.. the user in the link could still have a problem with his webrtc implementation.
Sounds weird to me that the codec will be present for one scenario but not to the other.
The best way to be sure is to make another test. Jitsi is an sfu, if I'll stream a h264 stream and then we'll know for sure..

**Comment by roeycohen** [comment]
ok, i've made some tests today.

- first i've tried using jitsi - you can force jitsi to use h264 using by adding options in the hash:
https://meet.jit.si/demotest#config.preferH264=true&config.p2p.enabled=false&config.disableSimulcast=true
h264 didn't work for me when using the above link at all.
when i've tried using they api directly it didn't work as well, unless i've tried with canary (70.0.3521.0).
so i guess jitsi have their own issues...
- next i've tried working with appr.tc. you can force it to use h264 using:
https://appr.tc/r/h264ant?vrc=h264&vsc=h264
i got the same results as in antmedia - it's works very nice in desktop, mobile (including canary) but in not webview.
i've even tried using canary as the webview, but got get the same results.

i've read in a few places, that sometimes the h264 doesn't work on all views.
https://github.com/jitsi/jicofo/issues/245#issuecomment-346079959
@mekya , is it something you can try to change and test?

I'll try to open a ticket at google, i hope they have some answers.
at least i've learnt a lot today...

**Comment by roeycohen** [comment]
seems the bug is already opened:
https://bugs.chromium.org/p/chromium/issues/detail?id=801501

**Comment by mekya** [comment]
I think, we can develop a cordova plugin that support H264 in WebRTC based on our already integrated WebRTC Android SDK. How do you think about that?
